content/crutchcorn/collections/art-of-accessibility/posts/art-of-a11y-semantic-markup/index.md

```diff
@@ -0,0 +1,486 @@
+
+> Keep in mind, we have to change these `aria` attributes on-the-fly as the information changes; say, with the `aria-selected` indicating which tab is active.
+>
+> HTML does not provide a way to automatically change the `aria` attributes for us without JavaScript.
+>
+> We'll build an interactive version of this `tab` component using React, Angular, and Vue later in this chapter that handles these things.
+
+While `role` is imperative in its usage here, it _can_ lead to subpar or even actively hostile user experiences for assistive technologies.
+
+This is because, using `role`, you have the ability to tell HTML that one element should be reflected to the end-user as an entirely different element, without actually providing any of the expected functionality.
+
+To explain this more, let's look at how an HTML `button` works.
+
+When you create an HTML element like `button`, the browser implicitly assigns it a`role` internally, regardless of if you assigned one or not.
+
+In this case:
+
+```html
+<button>Click me!</button>
+```
+
+Is implicitly treated by the browser as having `role="button"` assigned to it.
+
+> If that's the case, then surely `<div type="button">` must act the same as a `<button>`, right?
+
+Not quite.
+
+While you could create a partially analogous `button` element using a `div`:
+
+```html
+<div tabindex="0" role="button">Save</div>
+```
+
+ You might notice a problem with it when displayed on a web page:
+
+---
+
+<div tabindex="0" role="button">Save</div>
+
+---
+
+Notice that the fake "button" here doesn't appear to "press" down? There's no styling to indicate when the user is hovered over the "button", nor is there any visual indication when the user is hovered over the "button" with their mouse.
+
+This is why it's often **highly discouraged to use `role` in place of an HTML element with an implicit `role` enabled**; they simply don't have feature parity without a substantial amount of work and expertise.
+
+# Building a tab component with ARIA
+
+Now that we've seen a few examples of accessible, but non-interactive, markup let's see what we can do to breath life into these UI components using a framework.
+
+Namely, I want to demonstrate how we can build our own accessible tab component using aria attributes:
+
+![Two tabs of "JavaScript" and "Python" with JavaScript selected and a console.log inside](./styled_tabs.png)
+
+Let's start by reusing our markup from the previous section, and adding in some JavaScript to make the tabs interactive.
+
+<!-- ::start:tabs -->
+
+## React
+
+```jsx
+const App = () => {
+  const [activeTab, setActiveTab] = useState('javascript');
+
+  return (
+    <div>
+      <ul role="tablist">
+        <li
+          role="tab"
+          id="javascript-tab"
+          aria-selected={activeTab === 'javascript'}
+          aria-controls="javascript-panel"
+          onClick={() => setActiveTab('javascript')}
+        >
+          JavaScript
+        </li>
```

Comment:
These tablists are not keyboard accessible.

1. They are not in the tab order, well ideally one element in the tab order with managed or pseudo focus from arrow keys switching between them
2. `onClick` is only pointer click since these aren't buttons

Example: https://www.w3.org/WAI/ARIA/apg/patterns/tabs/examples/tabs-manual/

Reference: https://www.w3.org/WAI/ARIA/apg/patterns/tabs/

Reply:
Fixed the examples, now just need to rewrite the section themselves